google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,327 @@
+# Copyright 2017 The Forseti Security Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Rules engine for Bucket retention."""
+import collections
+from collections import namedtuple
+import itertools
+import threading
+
+from google.cloud.forseti.common.gcp_type import resource_util
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.common.util import relationship
+from google.cloud.forseti.scanner.audit import base_rules_engine as bre
+from google.cloud.forseti.scanner.audit import errors as audit_errors
+from google.cloud.forseti.services.inventory.base import resources
+
+
+LOGGER = logger.get_logger(__name__)
+
+VIOLATION_TYPE = 'RETENTION_VIOLATION'
+# Applyto.
+_APPLY_TO_BUCKETS = resources.GcsBucket.type()
+_APPLY_TO_RESOURCES = frozenset([_APPLY_TO_BUCKETS])
+
+
+class RetentionRulesEngine(bre.BaseRulesEngine):
+    """Rules engine for retention."""
+
+    def __init__(self, rules_file_path, snapshot_timestamp=None):
+        """Initialize.
+
+        Args:
+            rules_file_path (str): file location of rules
+            snapshot_timestamp (str): snapshot timestamp. Defaults to None.
+                If set, this will be the snapshot timestamp
+                used in the engine.
+        """
+        super(RetentionRulesEngine,
+              self).__init__(rules_file_path=rules_file_path)
+        self.rule_book = None
+
+    def build_rule_book(self, global_configs=None):
+        """Build RetentionRuleBook from the rules definition file.
+
+        Args:
+            global_configs (dict): Global configurations.
+        """
+        self.rule_book = RetentionRuleBook(self._load_rule_definitions())
+
+    def find_buckets_violations(self, buckets_lifecycle, force_rebuild=False):
```

Comment:
Does bucket need to have its own specific function? I think we could write a single function that works for all resources?